### PR TITLE
Add Oblivious HTTP (OHTTP) client for iOS [ci full]

### PR DESCRIPTION
**Pull request by moztcampbell**
Using the ohttp-rs and bhttp-rs Rust libraries to perform underlying encoding and encryption. This is exposed to Swift using URLRequest/Response types that are typical for the platform.

Note that only iOS is targetted since Gecko projects have a builtin OHTTP solution within Necko.

This use currently using the Rust Crypto libraries for underlying HPKE support, but this should be changed to NSS or OpenSSL in future.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

**Review by tarikeshaq — DISMISSED** [review]
This is awesome 🚀 🚀 !! I mostly had nits - generally, I'm happy with getting this out so iOS has their hands on it and we can iterate from there

Only requesting changes because we'll need to add this component to https://github.com/mozilla/application-services/blob/main/taskcluster/scripts/build-and-test-swift.py so the new taskcluster-based process can get this to iOS in the next nightly without any `rust-component-swift` changes at all. I'll update the docs to include that step!

 Once we do that, we can test it out by adding `[ci full]` in the PR title, then closing and re-opening the PR

**Comment by tarikeshaq** [comment]
For the [dependency check job](https://github.com/mozilla/application-services/actions/runs/5693210201/job/15432468078?pr=5749), we might need a fix up in https://github.com/mozilla/application-services/blob/9049484af9092b857fd701f411014b7aa7ac013a/tools/dependency_summary.py#L209 (there are some examples there)

Then running `./tools/regenerate_dependency_summaries.sh` and checking-in the changes the script does

File: components/as-ohttp-client/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,307 @@
+extern crate bhttp;
+extern crate ohttp;
+
+use std::collections::HashMap;
+use std::sync::Mutex;
```
[the remark this answers is not in the document]
Comment (moztcampbell) [reply components/as-ohttp-client/src/lib.rs:5]: Probably worth recommending parking_lot::Mutex in https://mozilla.github.io/uniffi-rs/udl/interfaces.html#concurrent-access as a starting point

File: components/as-ohttp-client/ios/ASOhttpClient/OhttpManager.swift [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        let data = try await fetchKey(url: gatewayConfigUrl)
+        Self.keyCache[gatewayConfigUrl] = (data, Date.now)
+
+        return data
+    }
+
+    private func invalidateKey() {
+        Self.keyCache.removeValue(forKey: URL(string: configUrl)!)
```
[the remark this answers is not in the document]
Comment (moztcampbell) [reply components/as-ohttp-client/ios/ASOhttpClient/OhttpManager.swift:59]: It looks like I can just force the caller of `OhttpManager::init` to pass URLs and that simplifies code too.

**Review by tarikeshaq — DISMISSED** [review]
A couple of tiny comments then we can land!

After fixing that comments tried to build Firefox iOS locally by running the taskcluster scripts and verified that the app builds and that it has access to `OhttpManager`, its constructor and the `data` function

**Review by tarikeshaq — DISMISSED** [review]
Awesome 🚀  thank you! 

I verified this builds okay with iOS and should be available for iOS engineers tomorrow (or we can expedite it to later today with a manual run!)

Do you mind adding a short entry to the changelog in https://github.com/mozilla/application-services/blob/main/CHANGELOG.md under the current (118) release?

(I'll also give you the permissions needed to land approved PR and so reviews don't go stale)

**Comment by moztcampbell** [comment]
Rebased onto v118 and added changelog entry.

**Review by tarikeshaq — APPROVED** [review]
🚢 🚢 🚢

**Pull request merged**